src/cpp/src/visual_language/qwen2vl/classes.hpp

```diff
@@ -64,13 +46,40 @@ class InputsEmbedderQwen2VL : public InputsEmbedder::IInputsEmbedder {
     bool prompt_has_image_tag(const std::string& prompt) const override;
 
 protected:
+    // A model for merging image embeddings (hidden states), rotary_pos_emb and attension_mask.
+    // Inputs:
+    //  - hidden_states: [N, embed_dim]
+    //  - rotary_pos_emb: [?, 40]
+    //  - attention_mask: [1, ?, ?]
+    // Output: [N, hidden_size]
+    std::unique_ptr<CircularBufferQueue<ov::InferRequest>> m_ireq_queue_vision_embeddings_merger;
+
+    ov::Tensor m_position_ids;
+    int64_t m_rope_delta = 0;
+    ov::Tensor m_merged_image_embeddings;
+
+    virtual ov::Tensor run_image_embeddings_merger(
+        const std::vector<EncodedImage>& images, 
+        const std::vector<size_t>& images_sequence, 
+        const size_t image_id);
+
     ov::Tensor merge_text_and_image_embeddings_qwen2vl(
         const ov::Tensor& input_ids,
         const ov::Tensor& text_embeds, 
         const ov::Tensor& merged_image_embeds,
         const int64_t image_pad_token_id
     );
 
+    std::pair<std::vector<ov::Tensor>, std::vector<std::array<size_t, 3>>> reorder_image_embeds_and_grid_thw(
+        const std::vector<EncodedImage>& encoded_images,
+        const std::vector<size_t>& images_sequence,
+        const size_t image_id
+    );
+
+    ov::Tensor get_attention_mask(const std::vector<std::array<size_t, 3>>& reordered_images_grid_thw);
+    
+    ov::Tensor concatenate_image_embeds(const std::vector<ov::Tensor> reordered_image_embeds);
```

Comment:
Move new functions outside of a class so it's easier to track which functions are overridden

Reply:
These new functions (`reorder_image_embeds_and_grid_thw`, `concatenate_image_embeds`, `get_attention_mask`, `get_rotary_pos_emb`) are reused in Qwen2_5_VL inherited class.

Reply:
Qwen2_5_VL can still call them if you declare them in src/cpp/src/visual_language/qwen2vl/classes.hpp

Reply:
Moved